src/platform/nxp/k32w/k32w0/KeyValueStoreManagerImpl.cpp

```diff
@@ -26,74 +26,232 @@
+            {
+                ChipLogProgress(DeviceLayer, "KVS, restored Matter key [%s] from flash with PDM id: %i", key_string_id,
+                                pdm_internal_id);
+            }
+        }
+    }
+
+    g_restored_from_flash = true;
+
+    return err;
+}
+
 CHIP_ERROR KeyValueStoreManagerImpl::_Get(const char * key, void * value, size_t value_size, size_t * read_bytes_size,
                                           size_t offset_bytes)
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
-    std::hash<std::string> hash_fn;
-    uint16_t key_id;
+    CHIP_ERROR err     = CHIP_DEVICE_ERROR_CONFIG_NOT_FOUND;
     uint8_t pdm_id_kvs = chip::DeviceLayer::Internal::K32WConfig::kPDMId_KVS;
-    size_t read_bytes;
+    std::unordered_map<std::string, uint8_t>::const_iterator it;
+    size_t read_bytes        = 0;
+    uint8_t key_id           = 0;
+    uint16_t pdm_internal_id = 0;
 
-    VerifyOrExit((key != NULL) && (value != NULL), err = CHIP_ERROR_INVALID_ARGUMENT);
-    key_id = hash_fn(key) % MAX_NO_OF_KEYS;
+    VerifyOrExit((key != NULL) && (value != NULL) && (RestoreFromFlash() == CHIP_NO_ERROR), err = CHIP_ERROR_INVALID_ARGUMENT);
 
-    ChipLogProgress(DeviceLayer, "KVS, get key id:: %i", key_id);
+    if ((it = g_kvs_map.find(key)) != g_kvs_map.end())
+    {
+        key_id          = it->second;
+        pdm_internal_id = chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id);
 
-    err = chip::DeviceLayer::Internal::K32WConfig::ReadConfigValueBin(
-        chip::DeviceLayer::Internal::K32WConfigKey(pdm_id_kvs, key_id), (uint8_t *) value, value_size, read_bytes);
+        err =
+            chip::DeviceLayer::Internal::K32WConfig::ReadConfigValueBin(pdm_internal_id, (uint8_t *) value, value_size, read_bytes);
+        *read_bytes_size = read_bytes;
 
-    *read_bytes_size = read_bytes;
+        ChipLogProgress(DeviceLayer, "KVS, get Matter key [%s] with PDM id: %i", key, pdm_internal_id);
+    }
 
 exit:
     return err;
 }
 
 CHIP_ERROR KeyValueStoreManagerImpl::_Put(const char * key, const void * value, size_t value_size)
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
-    std::hash<std::string> hash_fn;
-    uint16_t key_id;
+    CHIP_ERROR err = CHIP_ERROR_INVALID_ARGUMENT;
+    uint8_t key_id;
+    bool_t put_key     = false;
     uint8_t pdm_id_kvs = chip::DeviceLayer::Internal::K32WConfig::kPDMId_KVS;
+    std::unordered_map<std::string, uint8_t>::const_iterator it;
+    uint16_t pdm_internal_id = 0;
+
+    VerifyOrExit((key != NULL) && (value != NULL) && (RestoreFromFlash() == CHIP_NO_ERROR), err = CHIP_ERROR_INVALID_ARGUMENT);
+
+    if ((it = g_kvs_map.find(key)) == g_kvs_map.end()) /* new key */
+    {
+        for (key_id = 0; key_id < kMaxNumberOfKeys; key_id++)
+        {
+            std::set<uint8_t>::iterator iter = std::find(g_key_ids_set.begin(), g_key_ids_set.end(), key_id);
+
+            if (iter == g_key_ids_set.end())
+            {
+                assert(g_key_ids_set.size() < kMaxNumberOfKeys);
```

Comment:
```suggestion
                if (g_key_ids_set.size() >= kMaxNumberOfKeys)
                {
                    return CHIP_ERROR_NO_MEMORY;
                }
```

Reply:
Else we can get in a failure / reset loop, or (if assert is compiled out of the build) ignore the persistent storage bounds.

Assert shouldn't really be used for things that have a dependency on things from disk.